Initial add sonarqube-azuresql  -  Version 1 Beta

sstjean opened the pull request:
### Changelog
* v1.0 Beta
* Deploys and configures SonarQube
* Beta version does not have IIS Reverse Proxy for secure communications  between vm and user

### Description of the change

This template deploys an Azure SQL Server, Azure SQL DB, Windows Server 2012R2 VM (Standard DS1 v2) with SonarQube installed. This template can be deployed to a new resource group (recommended) or to an existing resource group.  

This Beta release deploys a SonarQube installation that is not secured. To secure the installation we need to implement an IIS Reverse Proxy on the SonarQube VM. This is planned for a future release, hopefully in the full v1.0 version. This means that your interaction with this SonarQube installation will happen in the clear over HTTP

azurecla commented:
Hi __@sstjean__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@sstjean__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

gatneil commented:
I think this is better named sonarqube-azuresql (without the 301) since sonaqube is not a Microsoft product

gatneil commented on 301-sonarqube-azuresql/README.md:
```diff
@@ -0,0 +1,107 @@
+This template deploys an Azure SQL Server, Azure SQL DB, Windows Server 2012R2 VM (Standard DS1 v2) with SonarQube installed.  This template 
+can be deployed to a new resource group (recommended) or to an existing resource group.  
+
+Once the deployment is complete you can increse the resources provided to SonarQube by changing the VM from a Standard DS1 v2 machine to any larger DS_x_ v2 machine without the need to redeploy SonarQube or migrate any data.
```
"increse" -> "increase"

sstjean replied:
fixed

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
```diff
@@ -0,0 +1,399 @@
+  "parameters": {
+    "sqVM_AppName": {
+      "type": "string",
+      "minLength": 1,
```
is there a good maxLength for this?

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
```diff
@@ -0,0 +1,399 @@
+    }
+  },
+  "variables": {
+    "GitHubRoot": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master",
```
please see the best practice about _artifactsLocation: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

gatneil replied:
this still needs to be resolved; it should be a parameter called _artifactsLocation instead of a variable

sstjean replied:
Yup.  I have a colleague fixing that part now. Hopefully will be in the PR this weekend.

sstjean replied:
The _artifactsLocation and _artifactsLocationSasToken parameters have been reinstated per the Contribution Guide.

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
do we not want to pass the connectionString and sqVmAdminPwd in protectedSettings?

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
```diff
@@ -0,0 +1,399 @@
+      "minLength": 1,
+      "maxLength": 32,
+      "metadata": {
+        "description": "The prefix of the public URL for the VM on the Internet. It should match with the following regular expression: ^[a-z][a-z0-9-]{1,61}[a-z0-9]$ or it will raise an error."
```
should probably also mention that it needs to be unique across all of Azure

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
please see best practice 9: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
this could break if we ever change the scheme for constructing FQDNs (which has happened before). Consider using the reference function like this: "value": "[concat('http://', reference(resourceId('Microsoft.Network/publicIPAddresses', variables('publicIPAddressName'))).dnsSettings.fqdn, '/')]",

gatneil commented on 301-sonarqube-azuresql/nested/azureDBDeploy.json:
```diff
@@ -0,0 +1,97 @@
+      "type": "Microsoft.Sql/servers",
+      "kind": "v12.0",
+      "location": "[resourceGroup().location]",
+      "apiVersion": "2014-04-01-preview",
```
is there a non-preview api version we can use?

sstjean replied:
This is the latest version available for Microsoft.Sql

gatneil replied:
ok

gatneil commented on 301-sonarqube-azuresql/nested/azureDBDeploy.json:
```diff
@@ -0,0 +1,97 @@
+          "name": "AllowAllWindowsAzureIps",
+          "type": "firewallrules",
+          "location": "[resourceGroup().location]",
+          "apiVersion": "2014-04-01-preview",
```
same

sstjean replied:
This is the latest version available for Microsoft.Sql

gatneil replied:
ok

gatneil commented on 301-sonarqube-azuresql/nested/azureDBDeploy.json:
```diff
@@ -0,0 +1,97 @@
+          "type": "databases",
+          "kind": "v12.0,user",
+          "location": "[resourceGroup().location]",
+          "apiVersion": "2014-04-01-preview",
```
same

sstjean replied:
This is the latest version available for Microsoft.Sql

gatneil replied:
ok

gatneil commented on 301-sonarqube-azuresql/nested/azureDBDeploy.json:
```diff
@@ -0,0 +1,97 @@
+  "outputs": {
+    "jdbcConnString": {
+      "type": "string",
+      "value": "[concat('jdbc:sqlserver://', reference(concat('Microsoft.Sql/servers/', parameters('sonarSqlName'))).fullyQualifiedDomainName, ':1433;database=', parameters('sonarDbName'), ';user=', parameters('adminLogin'), '@', parameters('sonarSqlName'), ';password=',parameters('adminPassword'))]"
```
not entirely sure how secure the "outputs" section is (could possibly show up in some log somewhere); you already called out that the template is insecure in the README, so this is fine, but just thought I'd bring it up for future improvements.

sstjean replied:
We are looking back into this.  We are going to let it go through for the insecure Beta but really want to tighten this up if possible.  I'll investigate whether we can return a SecureString to the main template.

gatneil replied:
ok

sstjean replied:
We found that a secure string can't be passed back so we are passing the connection string with a placeholder for the password.  Since the root template already has the password from the user, we can make the substitution before updating the config file.   This will be available in the next push to the branch.

gatneil commented:
Please address the above comments, then we can move forward with this PR :)

sstjean commented on 301-sonarqube-azuresql/azuredeploy.json:
We do want to change this and the team is tracking this as an unresolved bug.  We haven't had a chance to remediate it yet, but the problem may fall into the same category as returning the Connection String in plain text in the Output of the database template.  We definitely want this updated for the secured release, but for this Beta, can we move forward with this as-is under the "this is an insecure template" caveat?

sstjean commented on 301-sonarqube-azuresql/azuredeploy.json:
Fixed

sstjean commented on 301-sonarqube-azuresql/azuredeploy.json:
Fixed

sstjean commented:
Hi Niel - We've been working to update the template based on your feedback.  We still have a couple of outstanding issues we are working on (marked in their comments), but we wanted to get back to you sooner rather than later.

I wasn't able to reply to your comment to remove the "301-" from the name inline above, so I'm doing it here.  This template is being built by the Visual Studio ALM Rangers team under the guidance of Jean-Marc Prieur who is a PM in VSTS for Technical Debt.  Our understanding was that the 301- designation was a way to describe the relative complexity of the template.  I also see other non-Microsoft products like "301-drupal8-vmss-glusterfs-mysql" so I'm a bit confused by your earlier comment.  Could you elaborate please?

singhkays commented:
@gatneil We should look into removing 301 from the "301-drupal8-vmss-glusterfs-mysql" as that seems very application centric rather than VMSS concepts centric template. @sstjean The same would apply to this template as well since this is app centric.

sstjean commented:
@singhkays Ah, ok.  That makes sense then.  I'd like to recommend an update to the Contribution Guide reflecting the intended usage of the numeric prefixes.  I think that would have been helpful to us as we were building and naming this template.

sstjean commented:
@gatneil @singhkays I think we've covered all of your earlier comments and are ready for another review.  Thanks for your time.

gatneil commented on 301-sonarqube-azuresql/azuredeploy.json:
sure; this should be fine

gatneil commented:
Looks like the only issue left is the _artifactsLocation comment

gatneil commented on sonarqube-azuresql/README.md:
```diff
@@ -2,17 +2,17 @@
 
-[![Deploy Button](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Fsonarqube-azuresql%2Fazuredeploy.json)
-[![Visualize button](http://armviz.io/visualizebutton.png)](http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2Fsonarqube-azuresql%2Fazuredeploy.json)
+[![Deploy Button](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FALM-Rangers%2Fazure-quickstart-templates%2Fmaster%2Fsonarqube-azuresql%2Fazuredeploy.json)
```
this should point to Azure, not FALM-Rangers

sstjean replied:
Yup. You caught me. :)   I was testing a fix in the ALM-Rangers branch. I'll be sure to change it back to Azure in my final commit.

sstjean replied:
This has been reset to Azure in README.md and azuredeploy.json

sstjean commented:
@gatneil I believe we have addressed all of your concerns for this BETA.  I've just updated the path back to the Azure repo and reinstated the _artifactsLocation and _artifactsLocationSasToken parameters. Thanks for your time.
